zerver/views/message_fetch.py

```diff
@@ -801,20 +813,41 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
                          use_first_unread_anchor_val: bool=REQ('use_first_unread_anchor',
                                                                validator=check_bool, default=False),
                          client_gravatar: bool=REQ(validator=check_bool, default=False),
-                         apply_markdown: bool=REQ(validator=check_bool, default=True)) -> HttpResponse:
+                         apply_markdown: bool=REQ(validator=check_bool, default=True),
+                         is_web_public_query: bool=REQ(validator=check_bool, default=False)) -> HttpResponse:
+
     anchor = parse_anchor_value(anchor_val, use_first_unread_anchor_val)
     if num_before + num_after > MAX_MESSAGES_PER_FETCH:
         return json_error(_("Too many messages requested (maximum {}).").format(
             MAX_MESSAGES_PER_FETCH,
         ))
 
-    if user_profile.realm.email_address_visibility != Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+    if is_web_public_query:
+        # Web public queries are only allowed for limited types of narrows.
+        web_public_filter = {'operator': 'streams', 'operand': 'web-public'}
+        if narrow is None:
+            narrow = [web_public_filter]
+        elif not is_web_public_compatible(narrow):
+            return json_unauthorized()
+        else:
+            narrow.append(web_public_filter)
```

Comment:
We should probably extract this as a function, and unit-test it; I'd want to be sure that we handle multiple `streams` operators correctly.  (Ideally, we should just only end up with `streams:web-public`).